### PR TITLE
Radiology processing job endpoints

### PR DESCRIPTION
**Pull request by aauker**
Basic prototype done, new file  data_processing/services/radiology_service.py
* Accepts job input/config
* Submits/queues job in background

Screen Shot 2021-02-23 at 5 07 55 PM

Major changes:

 data_processing/common/Container.py
* Added configuration using configset/config.yaml, removed all hard-coded params
* If a connection to an object store is configured, files are uploaded during the save process

 data_processing/scanManager/*py
* Update using configset
* Generalize main function to accept params as a dictionary, adapt CLI tool to load a json and pass a dictonary

 data_processing/scanManager/config.py
* Added method to return configuration dictonary

### CODE REVIEW

File: data_processing/common/Container.py [hunk 1]
```diff
@@ -71,24 +74,42 @@ def __init__(self, params):
...
+        self._client = Minio(params['MINIO_URI'], access_key=params['MINIO_USER'], secret_key=params['MINIO_PASSWORD'], secure=False)
+        try:
+            for bucket in self._client.list_buckets():
+                self.logger.info("Found bucket %s", bucket.name )
+            self.logger.info("object_store_enabled=True")
+            params['object_store_enabled'] = True
+        except:
+            self.logger.warning("object_store_enabled=False")
+            params['object_store_enabled'] = False
+
```
Comment (doori) [inline data_processing/common/Container.py:95]: i would suggest to simply have `object_store_enabled` as a user input, with some default, and set up a client only when the param is true.

File: data_processing/scanManager/extractRadiomics.py [hunk 2]
```diff
@@ -18,30 +18,27 @@
 from data_processing.common.utils           import get_method_data
 from data_processing.common.Container       import Container
 from data_processing.common.Node            import Node
+from data_processing.common.config import ConfigSet
 
 # From radiology.common
 from data_processing.radiology.common.preprocess   import extract_radiomics
 
 logger = init_logger("extractRadiomics.log")
+cfg = ConfigSet("CONTAINER_CFG",  config_file="config.yaml")
```
Comment (doori) [inline data_processing/scanManager/extractRadiomics.py:27]: let's add a config.yaml.template if there isn't one already.

File: data_processing/services/radiology_service.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+from data_processing.common.config          import ConfigSet
+from data_processing.common.custom_logger   import init_logger
+
+from data_processing.scanManager.windowDicoms       import window_dicom_with_container
+from data_processing.scanManager.extractRadiomics   import extract_radiomics_with_container
+from data_processing.scanManager.generateScan       import generate_scan_with_container
+
+logger = init_logger("radiology-service.log")
+    
+cfg = ConfigSet("APP_CFG",  config_file="config.yaml")
```
Comment (doori) [inline data_processing/services/radiology_service.py:30]: please add an example app_config.yaml.template

File: data_processing/services/radiology_service.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+)
+
+# param models
+generate_scan_model = api.model("Generate Volumetric Image",
+    {
+        "job_tag": fields.String(description="Tag/name of output record", required=True, example='my_nrrd'),
+        "input_name": fields.String(description="Tag/name of input image record", required=True, example='dicoms'),
+        "file_ext": fields.String(description="A valid ITK image file extention", required=True, example='nrrd'),
+    }
+)
```
Comment (doori) [inline data_processing/services/radiology_service.py:69]: Great examples of the params!

File: data_processing/services/radiology_service.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+executor = ProcessPoolExecutor(NUM_PROCS) 
+
+# param models
+extract_radiomics_model = api.model("Extract Radiomics", 
+    {
+        "job_tag": fields.String(description="Tag/name of output record", required=True, example='my_radiomics'),
+        "image_input_name": fields.String(description="Tag/name of input image record", required=True, example='generated_mhd'),
+        "label_input_name": fields.String(description="Tag/name of label image record", required=True, example='user_segmentations'),
+        "enableAllImageTypes": fields.Boolean(description="Toggle all images", required=False, example=False),
+        "RadiomicsFeatureExtractor": fields.Raw(description="Key-value pairs for radiomics feature extractor", required=False, example={"interpolator": 'sitkBSpline'}),
```
Comment (doori) [inline data_processing/services/radiology_service.py:47]: do we want any defaults for Boolean and/or Raw fields?

File: data_processing/services/radiology_service.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        return f"Submitted job {job_id} with future {future}"
+
+@api.route('/mind/api/v1/generate_scan/<cohort_id>/<container_id>/submit', methods=['POST'])
+class API_generate_scan(Resource):
+    @api.expect(generate_scan_model, validate=True)
+    def post(self, cohort_id, container_id):
+        """Submit a generate scan Job"""
+        job_id = str(uuid.uuid4())
+        future = executor.submit (generate_scan_with_container, cohort_id, container_id, request.json)
+        return f"Submitted job {job_id} with future {future}"
```
Comment (doori) [inline data_processing/services/radiology_service.py:117]: It'd be good to standardize responses as a json with a status code and message/payload. 202 can be used as the status code for successful submission of async call.

File: data_processing/services/radiology_service.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+
+@api.route('/service/health', methods=['GET'])
+class API_heatlh(Resource):
+    def get(self):
+        try:
+            executor.submit(print, "Test")
```
Comment (doori) [inline data_processing/services/radiology_service.py:124]: Nice health check! do we need a pool submit here?

File: data_processing/services/radiology_service.py (hunk 6 above)
Comment (aauker) [reply data_processing/services/radiology_service.py]: Ah, agreed, changed now.

File: data_processing/services/radiology_service.py (hunk 7 above)
Comment (aauker) [reply data_processing/services/radiology_service.py:124]: Yes-- I encountered one case where the executor errors out and you can't submit jobs. This health check ensures the executor can accept submitted tasks.

**Review by doori — APPROVED** [review]
Looks great!

**Pull request merged**